radicle-node/src/service.rs

```diff
@@ -554,6 +567,38 @@ where
         }
     }
 
+    pub fn ping(&mut self, peer_ip: &net::IpAddr) -> Result<(), peer::SessionError> {
+        let peer = if let Some(peer) = self.sessions.get_mut(peer_ip) {
+            peer
+        } else {
+            return Err(SessionError::NotFound(*peer_ip));
+        };
+
+        let ping_upper = 16;
+        let ponglen = self.rng.u16(1..ping_upper);
+        let msg = Message::Ping {
+            ponglen,
+            zeroes: message::ZeroBytes(self.rng.u16(1..ping_upper)),
+        };
+        self.reactor.write(peer.addr, msg);
+
+        // How can I access the messages' ponglen field?
+        peer.ping = PingState::AwaitingResponse(ponglen);
+        Ok(())
+    }
+
+    pub fn ping_resp(&self, remote: &net::SocketAddr) -> Result<PingState, peer::SessionError> {
+        let peer_ip = remote.ip();
+        let peer = if let Some(peer) = self.sessions.get(&peer_ip) {
+            peer
+        } else {
+            return Err(SessionError::NotFound(remote.ip()));
+        };
+
+        // When should peer.ping be set to None again?
+        Ok(peer.ping)
+    }
```

Comment:
Looks like this function just exists to get the ping state? It should be removed in favor of accessing the session directly, if it's only for testing.